pkg/serviceaccounttoken/secret.go

```diff
@@ -204,29 +233,89 @@ func isSecretForServiceAccount(secret *corev1.Secret, sa *corev1.ServiceAccount)
 	return sa.Name == annotation
 }
 
-func createServiceAccountSecret(ctx context.Context, sa *corev1.ServiceAccount, secretLister secretLister, secretClient clientv1.SecretInterface, lockKey string) (*corev1.Secret, error) {
-	mutex := getLock(lockKey)
-	mutex.Lock()
-	defer func(key string) {
-		mutex.Unlock()
-		lockMap.Delete(lockKey)
-	}(lockKey)
-
-	// We could have been waiting for the Mutex to unlock in a parallel run of
-	// createServiceAccountSecret - check again for the secret existing.
-	secret, err := ServiceAccountSecret(ctx, sa, secretLister, secretClient)
+func createServiceAccountSecret(ctx context.Context, sa *corev1.ServiceAccount, secretLister secretLister, secretClient clientv1.SecretInterface) (*corev1.Secret, error) {
+	sc := SecretTemplate(sa)
+	secret, err := secretClient.Create(ctx, sc, metav1.CreateOptions{})
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("error ensuring secret for service account %s: %w", logKeyFromObject(sa), err)
 	}
-	if secret != nil {
-		return secret, nil
+
+	return secret, nil
+}
+
+// returns true if the secret has changed.
+func annotateSAWithSecret(ctx context.Context, sa *corev1.ServiceAccount, secret *corev1.Secret, saClient clientv1.ServiceAccountInterface, secretClient clientv1.SecretInterface) (*corev1.ServiceAccount, bool, error) {
+	if sa.Annotations == nil {
+		sa.Annotations = map[string]string{}
 	}
+	secretAnnotation := secret.Namespace + "/" + secret.Name
 
-	sc := SecretTemplate(sa)
-	secret, err = secretClient.Create(ctx, sc, metav1.CreateOptions{})
+	// If the SA is already annotated with the secret already
+	if ann := sa.Annotations[ServiceAccountSecretRefAnnotation]; ann == secretAnnotation {
+		return sa, false, nil
+	}
+
+	sa.Annotations[ServiceAccountSecretRefAnnotation] = secretAnnotation
+
+	updated, err := saClient.Update(ctx, sa, metav1.UpdateOptions{})
+	if err == nil {
+		return updated, false, nil
+	}
+	if !apierrors.IsConflict(err) {
+		logrus.Debugf("Failed to update ServiceAccount for %s: %s", logKeyFromObject(sa), err)
+		return nil, false, err
+	}
+
+	// Rollback the optimistically created secret
+	logrus.Debugf("Rolling back ServiceAccount secret for %s", logKeyFromObject(secret))
+	if err := secretClient.Delete(ctx, secret.Name, metav1.DeleteOptions{}); err != nil {
```

Comment:
I think that this error case could used further consideration. Since we don't cleanup after the annotation is set, this could lead to us having a few duplicates (though it's a fixed amount once we get the annotation set, we won't make any more). 

We can address in a future PR if you want since I agree that this is overall better.

Reply:
I think we can do this _after_ we've seen how much of an impact this has.